### ISSUE TITLE
Half a variadic tuple rest sandwich matching algorithm missing

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#49105]
Half a variadic tuple rest sandwich missing

# Bug Report
### 🔎 Search Terms

variadic tuple assignment matching algorithm

#### Background

When a non-generic variadic tuple of arbitrary depth is declared, 
it is transformed into one of two possible final flat canonical forms,
as described by this comment
```
// We have optional, rest, or variadic elements that may need normalizing. Normalization ensures that all variadic
// elements are generic and that the tuple type has one of the following layouts, disregarding variadic elements:
// (1) Zero or more required elements, followed by zero or more optional elements, followed by zero or one rest element.
// (2) Zero or more required elements, followed by a rest element, followed by zero or more required elements.
// In either layout, zero or more generic variadic elements may be present at any location.
```
found in `checker.ts`, `function createNormalizedTupleType`.

#### Bug 

However, the current unidirectional matching algorithm in `checkArrayLiteral` does not work correctly with the output from 
`createNormalizedTupleType`.  It incorrectly gives errors.  (Incorrect meaning, it returns errors for lhs values that could pass if a full combinatorial search were used.

That's because `checkArrayLiteral` performs a unidirection (left to right) search only, whereas the output from `createNormalizedTupleType` requires a bidirectional search. 

One solution would be be to change the output of `createNormalizedTupleType` so that it all elements to the right off the first rest element were rolled into a single rest.  Then the unidirectional search could be used with returning false errors.

However, a bidirectional search takes the same O(1) time as the unidirectional search, yet 
yet produces fewer false postives (false passes) than the unidirectional search.  Therefore the bidirectional search is prefereable.

The bidirectional matching algorithm intended by the output of `createNormalizedTupleType` is 

- match values from the front until the rest element is reached, 
- match values from the back until the rest element is reached,
- match the remaining values to the rest element.

### 🕗 Version & Regression Information

The example wrong assignment results for variadic tuples are stable in version range 4.2.3 ~ 4.8 dev.

### ⏯ Playground Link

[playground](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=false#code/PTAEAEGcBcCcEsDG0Bco4FcCmAoEEsAPAQ2QHkAHaeAewDtiAbABVhoq1mgE8AVbjpDQAzJpFx4AVJJyhJoXgAssoAEZZhNWFjQAKMgGlgAUVhtYASlAANULqIdkWACbAMdB1ifOrdGtGJqekhQenRlUFhFENl5GmFwlWJISHgAczoAWyw6aBDibVAANyZ4Z1BNWGLOVODQABYAOgAmRoBmUAA-BsaADlBnLCLG2OAcHABvWVAZ-H4OBWbQAEkAZVAAOTJeUGJIrBhQSGI6ZwB3JEUKrUzA0ABaXcYz4m4Qv1BERRO0lXg6faQDCMaBqLDQM5YHJgypJU67YTQTjTGY8Ba8ACMoAAvKAANqEFB0DCZdSwAA03AA-ChVDQaIwsCcALoAbhR6AEKl4S1xeMaAsx5NAACIACQitmzMAAcQAqssjooaGcQmjubz8cTSZxhXSGUyAQAfUDuQbCf4uYXiyWfE70JBMTkLLGVW6g3QfbQwCzjGYzRDBUHQZrEFA8nH4jHWiVS-ATGFaHSgUzmOyebwWYXERGcNCprTpwiOJE+UAAXw5gbohxDqnDmrx0fQsGwMclrNA8cT2jQhiLJZcWYRSNgfYMA68paslcrkw5cy5CgxSzWu0BoOOpwuX2usHdD3C8BCRGImQojKVKpC3uBeXCd3g57YJVUjG4A3gwmEnDBEKhALqLCuzwjmo4jP6zoqAAMskoK4ja7KQeqy5YnyhIigAeiK5ICo03C0vSjInHizJshyKGYhimSRniABeNIwAgdBpORyFLpijZ4VRwrcRi1HkjacZgJRK60VhOGgHhuhMf8aSgCa+rEcapqnBolo+KR7bMnafh0I6jBQaASxuncujepuJznJcvoLtIoAcvIywAtAEQeIQwZnDQuwpOkWQ5KCp7noyQiOf6QFJmgrkqGkjAYNQfgYDeByBYy2S5KAAAGnFNsymWgGk2guPA76gEy97RZE0RZWKmXChQ2hFAFclZTlzR5RUbA0YGZhTqV7pfC1UQhJltWNA5kHyDMYF5hU8CwIc2UrniAAMHUDcoN7VRJwoWWVaUBa1y0YutgSDSxoDHgMGg-tozjCktzR4u1mUbQciSgIwcGihKwphZBVlzQwjD9Wdm0fXtWAHRlj25ZlEH+pIYyQdWtYrmGnHidh2mdt2EW9imZiFro7gZtO2a5mOoD9vYxZTkOFYTf6qPBiu9aY3yO0tm2P0dl2YAJvjKgoAWVQk+5g4+BTo7jpOmYVn6-r4M5H3uZ53nJKkGTpYFJDBQcwqVULqlky4ZVEwtuyFA1QwBWbqgfg4jBIPAPB2jAQ3bRK6DeZVzujk6KGu-sTClVd2QnMG3nqOSC5gGciiXJd7z+J8NDnoE8BvioFyuRDKX7VgOtHU9J3wzMShXVdABWSVR5dIN17AgQqMqZwfe4ZTzV4rv0E60E+6AABKoAbS1V1ebAADWclxyB5SICcRzp+CT7vb8wYRAgaSKKCJyQJCsAIwGQboCuiDhmJnPY7zPmgLBMDCaAgsaJF1MTrTkvDjNVM06bZaVhRqfaAK5nCX0bFzTAWB2x3wftAJ+L9YSy0-vTKWI5Zp-zpvLWcOAgA)

### 💻 Code

```ts
// @strict: true
// @exactOptionalPropertyTypes: false

/**
 * The before: (OK/Error) X (expected/unexpected) notations on the rhs 
 * of the assignments are valid for versions 4.2.3 ~ 4.8 dev.
 */

{
    // Type T2 IS NOT a rest sandwich format - always no change in result between before and after
    type T1 = [x:number,y?:boolean];
    type T2 = [...T1, "$"];  // GUI shows type T2 = [number, boolean | undefined, "$"] canonical type 1 format (no rest)

    const t2a:T2 = [1, "$"]; // { before: Error (expected), after: Error (expected) }  
    const t2b:T2 = [1, true, "$"]; // { before: OK (expected), after: OK (expected) }
}

{
    // Type T12 IS a rest sandwich format - this example shows results that improvably differ between before and after.
    type Last = "$";
    type T11 = [x:"^",...y:boolean[]];
    type T11m = [z?:string];
    type T12 = [...T11, ...T11m,"$"]; // type T12 = ["^", ...(string | boolean | undefined)[], "$"] canonical type 2 format (rest sandwich)

    /** 
    * In the next two assignment examples:
    *   before: the glutinous rest element `T12[1]` greedily eats the rhs `$`, preventing `T12[2]` from correctly matching rhs `$`. 
    *   after: first `T12[0]` matches rhs "^", rest element `T12[1]` matching is deffered, `T12[2]`matches the last "$", 
    *     and finally matches the rest element `T12[1]`.
    */
    const t12a:T12 = ["^", "$"]; // { before: Error (unexpected), after: OK (expected) }         // BUG
    const t12b:T12 = ["^", true, "$"]; // { before :Error (unexpected), after: OK (expected) } // BUG

    // In the next two assignment examples, the before unexpected errors are prevented by explicity casting rhs "$" to the literal type it really is meant to be,
    // which is not compatible with the rest element `T12[1]`.  This is just to illustrate how the unidirecitonal L to R matching is working
    // and can sometimes get the right answer.
    const t12c:T12 = ["^", "$" as Last]; // { before: OK (expected), after: OK (expected) }
    const t12d:T12 = ["^", true, "$" as Last]; // { before: OK (expected), after: OK (expected) }
}
```

### 🙁 Actual behavior

Matching from left to right (Alg1).
Invalid results found.

### 🙂 Expected behavior

Using the bidirectional, rest-centered search described above no invalid results would be found.

### Deep analysis

In order to properly analyze the the specs, I felt I had to write an implementation of the bidirectional search
and run it on the baseline tests.  That's work is in the "pull draft" #49106.
Of course I assume this "pull draft" will not be used, but the analysis result are still valid, and potentially useful. 
 
To summarize the analysis, the way the code is currently written (4.7) there is assumption that array context types 
don't necessarily need to expose a length for use in matching.  However that won't necessarily work when
trying to match with a bidirectional search. Might be able to get away with negative indices 
(distance from the end) and still avoid length.  But stretching a rest element to infinite length won't work when a bidirectional search is required.

**Comment by fatcerberus** [comment]
> half a variadic tuple rest sandwich

Doesn’t sound very tasty.  Who ate the other half?

**Comment by craigphicks** [comment]
@fatcerberus - "There's nobody here but us mice!"

**Comment by craigphicks** [comment]
I'm going to report a simpler version of the bug

**Comment by craigphicks** [comment]
Reopening after the topic arose again.

**Comment by craigphicks** [comment]
See the newer version #49138.
